src/EntityComponentManager.cc

```diff
@@ -270,6 +276,154 @@ Entity EntityComponentManagerPrivate::CreateEntityImplementation(Entity _entity)
+        << "] with a name of [" << _name << "], but another entity already "
+        << "has this name." << std::endl;
+      return kNullEntity;
+    }
+    uniqueNameGenerated = true;
+  }
+
+  auto clonedEntity = this->CreateEntity();
+
+  if (_parent != kNullEntity)
+  {
+    this->SetParentEntity(clonedEntity, _parent);
+    this->CreateComponent(clonedEntity, components::ParentEntity(_parent));
+  }
+
+  // make sure that the cloned entity has a unique name
+  auto clonedName = _name;
+  if (!uniqueNameGenerated)
+  {
+    if (clonedName.empty())
+    {
+      auto originalNameComp = this->Component<components::Name>(_entity);
+      clonedName =
+        originalNameComp ? originalNameComp->Data() : "cloned_entity";
+    }
+    uint64_t suffix = 1;
+    while (kNullEntity != this->EntityByComponents(
+          components::Name(clonedName + "_" + std::to_string(suffix))))
+      suffix++;
+    clonedName += "_" + std::to_string(suffix);
+  }
+  this->CreateComponent(clonedEntity, components::Name(clonedName));
+
+  // copy all components from _entity to clonedEntity
+  for (const auto &type : this->ComponentTypes(_entity))
+  {
+    // skip the Name and ParentEntity components since those were already
+    // handled above
+    if ((type == components::Name::typeId) ||
+        (type == components::ParentEntity::typeId))
+      continue;
+
+    auto originalComp = this->ComponentImplementation(_entity, type);
+    auto clonedComp = originalComp->Clone();
+
+    if (type == components::Light::typeId)
+    {
+      // set the sdf::Light's name to clonedName
+      auto derivedComp =
+        static_cast<components::Light *>(clonedComp.get());
+      derivedComp->Data().SetName(clonedName);
+    }
+    else if (type == components::Actor::typeId)
+    {
+      // set the sdf::Actor's name to clonedName
+      auto derivedComp =
+        static_cast<components::Actor *>(clonedComp.get());
+      derivedComp->Data().SetName(clonedName);
+    }
+
+    this->CreateComponentImplementation(clonedEntity, type, clonedComp.get());
+  }
+
+  // keep track of canonical link information (for clones of models, the cloned
+  // model should not share the same canonical link as the original model)
+  if (auto modelCanonLinkComp =
+      this->Component<components::ModelCanonicalLink>(clonedEntity))
+  {
+    // we're cloning a model, so we map the cloned model to the original
+    // model's canonical link
+    _oldModelCanonicalLink[clonedEntity] = modelCanonLinkComp->Data();
+  }
+  else if (this->Component<components::CanonicalLink>(clonedEntity))
+  {
+    // we're cloning a canonical link, so we map the original canonical link
+    // to the cloned canonical link
+    _oldToClonedCanonicalLink[_entity] = clonedEntity;
+  }
+
+  for (const auto &childEntity :
+      this->EntitiesByComponents(components::ParentEntity(_entity)))
+  {
+    this->Clone(childEntity, clonedEntity, "", true, _oldModelCanonicalLink,
+        _oldToClonedCanonicalLink);
```

Comment:
I think we should check the return value of this function to handle children not being created correctly.

Reply:
https://github.com/ignitionrobotics/ign-gazebo/pull/959/commits/ba5f116ae6dbcc4602a15853cec2dae1d9d7aec6